pkg/scheduler/metrics/resources/resources.go

```diff
@@ -0,0 +1,296 @@
+// and phase of lifecycle. Each pod returns two series per resource - one for
+// their aggregate usage (required to schedule) and one for their phase specific
+// usage. This allows admins to assess the cost per resource at different phases
+// of startup and compare to actual resource usage.
+func NewPodResourcesMetricsCollector(podLister corelisters.PodLister) metrics.StableCollector {
+	return &podResourceCollector{
+		lister: podLister,
+	}
+}
+
+type podResourceCollector struct {
+	metrics.BaseStableCollector
+	lister corelisters.PodLister
+}
+
+func (c *podResourceCollector) DescribeWithStability(ch chan<- *metrics.Desc) {
+	podResourceDesc.Describe(ch)
+}
+
+func (c *podResourceCollector) CollectWithStability(ch chan<- metrics.Metric) {
+	pods, err := c.lister.List(labels.Everything())
+	if err != nil {
+		return
+	}
+	reuseReqs, reuseLimits := make(v1.ResourceList, 4), make(v1.ResourceList, 4)
+	for _, p := range pods {
+		_, reqs, _, limits, _, terminal := podRequestsAndLimitsByLifecycle(p, reuseReqs, reuseLimits)
+		if terminal {
+			// terminal pods are excluded from resource usage calculations
+			continue
+		}
+		for _, t := range []struct {
+			desc    resourceLifecycleDescriptors
+			total   v1.ResourceList
+			running v1.ResourceList
+		}{
+			{
+				desc:  podResourceDesc.requests,
+				total: reqs,
+			},
+			{
+				desc:  podResourceDesc.limits,
+				total: limits,
+			},
+		} {
+			for resourceName, val := range t.total {
+				var unitName string
+				switch resourceName {
+				case v1.ResourceCPU:
+					unitName = "cores"
+				case v1.ResourceMemory:
+					unitName = "bytes"
+				case v1.ResourceStorage:
+					unitName = "bytes"
+				case v1.ResourceEphemeralStorage:
+					unitName = "bytes"
+				default:
+					switch {
+					case isHugePageResourceName(resourceName):
+						unitName = "bytes"
+					case isAttachableVolumeResourceName(resourceName):
+						unitName = "integer"
+					}
+				}
+				var priority string
+				if p.Spec.Priority != nil {
+					priority = strconv.FormatInt(int64(*p.Spec.Priority), 10)
+				}
+				recordMetricWithUnit(ch, t.desc.total, resourceName, unitName, val, p.Namespace, p.Name, p.Spec.NodeName, p.Spec.SchedulerName, priority)
+			}
+		}
+	}
+}
+
+func recordMetricWithUnit(
+	ch chan<- metrics.Metric,
+	desc *metrics.Desc,
+	resourceName v1.ResourceName,
+	unit string,
+	val resource.Quantity,
+	namespace, name, nodeName, schedulerName, priority string,
+) {
+	if val.IsZero() {
+		return
+	}
+	ch <- metrics.NewLazyConstMetric(desc, metrics.GaugeValue,
+		val.AsApproximateFloat64(),
+		namespace, name, nodeName, schedulerName, priority, string(resourceName), unit,
+	)
+}
+
+// addResourceList adds the resources in newList to list
+func addResourceList(list, newList v1.ResourceList) {
+	for name, quantity := range newList {
+		if value, ok := list[name]; !ok {
+			list[name] = quantity.DeepCopy()
+		} else {
+			value.Add(quantity)
+			list[name] = value
```

Comment:
Just `list[name].Add(quantity)`?

Reply:
I believe that would be illegal because it's not a pointer receiver.